### DISCUSSION TITLE
Problems with ShellStream

### DISCUSSION THREAD

**Discussion by IgorMilavec** [sshnet/SSH.NET#926] [Ideas]
This initial comment will be updated with all the identified issues with ShellStream:

Defects:

1.  Read(byte[], ini, int) is not blocking when no data is available (#63, #923)
1. ShellStream object is disposed when the channel is closed (#424, #453, #672, #917)
1. Expect hangs if the channel is closed (#670)
1. Flush() should be called automatically in Write*() (#180, #301)
1. ShellStream.Expect returns the entire input buffer instead of matched text (#140, #454, #833)
1. ShellStream.Expect does not handle multibyte chars properly (#383)

Enhancements:
1. Add Closed event (#44, #432)
1. Recognize Lf line ending in addition to CrLf (#748)
1. Timeouts are not observed properly in Expect and ReadLine (#365, #714)
1. Flush() should be called automatically in Read() and Write*() (#302, #444)
1. Write(string and WriteLine(string) do not use write buffer (#301)
1. Add ability to change window size (#40)
1. Inefficient buffer implementation (#793, #800)

Requested examples:
 - How to run a background process (&) (#16)

**Comment by IgorMilavec** [comment]
I went through all the open issues and tried to summarize all of the problems with ShellStream. I you have something to add or discuss, please comment below and I will edit the first comment so we have a summarized description.

**Comment by IgorMilavec** [comment]
Added #432, #793 and #800.

**Reply by PatrickSchmidtSE** [reply]
Thank you for that :)

**Comment by IgorMilavec** [comment]
OK, so obviously we have a lot to do with ShellStream... IMO bugs should be fixed ASAP, but there is a large number of them, so fixing them one at a time might take a long time. Due to partial interdependence of some of the bugs, handling them in separate PRs would also create merge conflicts. @drieseng I know you prefer small focused PRs, but would you be willing to accept batched fixes in this case?

As tests did not catch these defects, we should first add a (failing) test for each of these scenarios before fixing them.

As a general design note it seems to me that a lot of problems originate from the ShellStream providing too much functionality. A lot of functionality is duplicated with StreamReader (which will properly handle character encoding). Expect() seems to be a nice utility for simple scenarios, but fails or is inefficient in more complex scenarios. I know I'm discussing breaking changes here, so we should thread carefully, but I really feel that with proper class design SSH.Net could be more stable and more performant. And we don't need to do breaking changes over night, [Obsolete] is a great tool in the long run.

**Reply by drieseng** [reply]
@IgorMilavec I propose to with an issue or discussion that we'll use to identify the bugs and any proposed changes.
We can then see if we group multiple fixes or changes in a single PR.
Works for you?

**Update**:
Perhaps I should read the description of the issue  :-) LOL

**Reply by drieseng** [reply]
@IgorMilavec I'd start with fixing the first defect. I'd also like to add some integration tests for **ShellStream** (to https://github.com/sshnet/IntegrationTests).

Let's also discuss what we want to do for the second defect. I see the following scenario's:
* the server closes the channel (shell)
* the server closes the SSH session.
* the client closes the channel (shell). eg. disposes the **ShellStream**.
* the client closes the SSH session (eg. disposes the **SshClient**).

Do we want to expose an event? Should the event allow one to distinguish the exact scenario?
Should the event also fire if the client (user) closes the channel or SSH session?

**Comment by VAllens** [comment]
That's great.

**Comment by IgorMilavec** [comment]
I've created #923 for the first defect.

Here is the behavior I would expect in these cases:
 - the server closes the channel (shell): Read returns 0, Write returns 0
 - the server closes the SSH session: Same as above
 - the client closes the channel (shell). eg. disposes the ShellStream.: SImple: ObjectDisposedException
 - the client closes the SSH session (eg. disposes the SshClient): this is the tricky one... the same behavior one gets when creating a NetworkStream from a Socket and then closes the socket?

So, generally, I would like ShellStream to behave the way dotnet built-in streams behave and the way every developer intuitively expects it to behave.

As for the events: at this time I'd rather not add any new API surface, unless someone shows he needs them and there is no other way to do it.

**Comment by WojciechNagorski** [comment]
@IgorMilavec Do you want to continue your work on ShellStream? I intend to include all ShellStream related tasks in the next release.

FYI @Rob-Hague

**Comment by jscarle** [comment]
@WojciechNagorski There appears to be only 4 issues left that are linked to this discussion.

- I think #16 can now be considered a stale discussion. It mostly seems to be a user discussion and considering this was 8 years ago, I think it's safe to close this issue.
- #40 seems like a reasonable request, but I believe this should addressed in a future release.
- As for #44, I need your opinion or @Rob-Hague 's opinion. I don't know what the current thoughts or plans are in terms of events. I know that .NET in general seems to be moving away from events and replacing them with `Action` and `Func`, so I'm not sure if this should be planned or not.
- I'm not sure we'd want to do #424. `IsDisposed` seems to [mostly exist](https://www.google.com/search?q=site%3Alearn.microsoft.com+c%23+%22IsDisposed%22) in concepts related to the UI or direct ties to unmanaged ressources. So again, is this something we still want to do?

These issues are generally very old and haven't really generated repeated discussions or issues. Most of these may no longer apply with all of the recent internal changes.

My proposal would be to close #16, #44, #424 and keep only #40 open but planned for a future release.

**Comment by WojciechNagorski** [comment]
I keep open only https://github.com/sshnet/SSH.NET/issues/40 
@jscarle Thanks for your summary!

**Reply by scott-xu** [reply]
I think https://github.com/sshnet/SSH.NET/issues/40 would get fixed once https://github.com/sshnet/SSH.NET/pull/1062 is ready and merged. @glenkleidon
